nemo/utils/config_utils.py

```diff
@@ -0,0 +1,170 @@
+
+    # Update optional configs
+    model_cls = _update_subconfig(
+        model_cls, update_cfg, subconfig_key='train_ds', drop_missing_subconfigs=drop_missing_subconfigs
+    )
+    model_cls = _update_subconfig(
+        model_cls, update_cfg, subconfig_key='validation_ds', drop_missing_subconfigs=drop_missing_subconfigs
+    )
+    model_cls = _update_subconfig(
+        model_cls, update_cfg, subconfig_key='test_ds', drop_missing_subconfigs=drop_missing_subconfigs
+    )
+    model_cls = _update_subconfig(
+        model_cls, update_cfg, subconfig_key='optim', drop_missing_subconfigs=drop_missing_subconfigs
+    )
+
+    # Add optim and sched additional keys to model cls
+    model_cls = _add_subconfig_keys(model_cls, update_cfg, subconfig_key='optim')
+
+    # Perform full merge of model config class and update config
+    # Remove ModelPT artifact `target`
+    if 'target' in update_cfg.model:
+        # Assume artifact from ModelPT and pop
+        if 'target' not in model_cls.model:
+            with open_dict(update_cfg.model):
+                update_cfg.model.pop('target')
+
+    model_cfg = OmegaConf.merge(model_cls, update_cfg)
+
+    return model_cfg
+
+
+def _update_subconfig(
+    model_cfg: DictConfig, update_cfg: DictConfig, subconfig_key: str, drop_missing_subconfigs: bool
+):
+    """
+    Updates the ModelPTConfig DictConfig such that:
+    1)  If the sub-config key exists in the `update_cfg`, but does not exist in ModelPT config:
+        - Add the sub-config from update_cfg to ModelPT config
+
+    2) If the sub-config key does not exist in `update_cfg`, but exists in ModelPT config:
+        - Remove the sub-config from the ModelPT config; iff the `drop_missing_subconfigs` flag is set.
+
+    Args:
+        model_cfg: A DictConfig instantiated from the ModelPTConfig subclass.
+        update_cfg: A DictConfig that mirrors the structure of `model_cfg`, used to update its default values.
+        subconfig_key: A str key used to check and update the sub-config.
+        drop_missing_subconfigs: A bool flag, whether to allow deletion of the ModelPTConfig sub-config,
+            if its mirror sub-config does not exist in the `update_cfg`.
+
+    Returns:
+        The updated DictConfig for the ModelPTConfig
+    """
+    with open_dict(model_cfg.model):
+        # If update config has the key, but model cfg doesnt have the key
+        # Add the update cfg subconfig to the model cfg
+        if subconfig_key in update_cfg.model and subconfig_key not in model_cfg.model:
+            model_cfg.model[subconfig_key] = update_cfg.model[subconfig_key]
+
+        # If update config does not the key, but model cfg has the key
+        # Remove the model cfg subconfig in order to match layout of update cfg
+        if subconfig_key not in update_cfg.model and subconfig_key in model_cfg.model:
+            if drop_missing_subconfigs:
+                model_cfg.model.pop(subconfig_key)
```

Comment:
Wait, why would we want to do this. I thought the point of dataclasses was to enforce a structure for the yaml.

Reply:
nvm